### PR TITLE
feat(window): dispatch the `installing` event

### PR DESCRIPTION
**Pull request by Doctor-wu**
support `installing` callback
issue for workbox: https://github.com/GoogleChrome/workbox/issues/3123

### CODE REVIEW

**Comment by userquin** [comment]
The title should be changed to `feat: fire installing and adding updatefound client events`, I think we should fire both events being external or not: dispatching the event should be outside the else statament (below the if/else).

We also need to add proper documentation for both events.

I'm doing some tests in my local using [workbox-window-es](https://github.com/userquin/workbox-window-es) (I'm going to deprecate it in favor of this one and change pwa plugin to use `serwist`).

File: packages/window/src/Serwist.ts [hunk 1]
```diff
@@ -386,15 +387,22 @@ export class Serwist extends SerwistEventTarget {
...
       if (process.env.NODE_ENV !== "production") {
         if (navigator.serviceWorker.controller) {
           logger.log("Updated service worker found. Installing now...");
         } else {
           logger.log("Service worker is installing...");
         }
       }
+      // dispatch the `installing` event when the SW is installing.
+      this.dispatchEvent(
+        new SerwistEvent("installing", {
```
Comment (userquin) [inline packages/window/src/Serwist.ts:399]: I think we should fire the event even it is external

**Comment by DuCanhGH** [comment]
@Doctor-wu waiting for the PR to be set to Ready :)

**Comment by Doctor-wu** [comment]
> @Doctor-wu waiting for the PR to be set to Ready :)

We need to do some tests, will set it ready ASAP

**Comment by DuCanhGH** [comment]
@Doctor-wu working on my end for the non-external service worker :)
image
I think this should work for external service workers as well, but yeah, it is better to check.

**Comment by Doctor-wu** [comment]
> @Doctor-wu working on my end for the non-external service worker :)
> image
> I think this should work for external service workers as well, but yeah, it is better to check.
> 
> Edit: ah, actually, just noticed `ServiceWorker.state` is "activated", and I think that's not supposed to be the case...
> 
> Edit 2: the logic doesn't seem wrong though, so I guess we can brush that off as... the service worker changing its state from "installing" to "activated" while `_onUpdateFound` was processing? Thinking if we can do the dispatch before the if-else branching currently above though.

@DuCanhGH Agree with the state was changed when the process were going down. It appears to almost the same if we lift the dispatch above the if/else statement.

**Comment by Doctor-wu** [comment]
Works well for external service worker~

**Comment by DuCanhGH** [comment]
@Doctor-wu your call on when to set the PR to Ready :)

**Comment by DuCanhGH** [comment]
@Doctor-wu thanks for the PR! Only after publishing 8.2.0 did I realize that your PR was already set to Ready 😅 Perhaps you will have to wait for a day or two to see your PR live, for publishing two minor versions within a day seems weird. That would give me a bit of time to support development mode in `@serwist/vite` as well.

**Comment by Doctor-wu** [comment]
@DuCanhGH That's OK, wait for your release~

**Pull request merged**